### DISCUSSION TITLE
What storage format does/will Outline use?

### DISCUSSION THREAD

**Discussion by matthijskooijman** [outline/outline#7396] [General] [answered]
I'm currently playing around with Outline and I am a bit confused about the storage format used internally for storing documents.

At first glance, it seems like you are editing markdown, but the editor is really a rich-text WYSIWYG editor that supports markdown syntax as input (which is nice!).

However, when looking at the `documents.info` (retrieve document) API call, the document body does look like it is still Markdown (possibly with extensions?), which is also what the [API docs suggest](https://www.getoutline.com/developers#tag/documents)  ("Documents are stored in [Markdown](https://spec.commonmark.org/) formatting"). 

However, in various discussions about wanting a raw markdown editing mode, I see statements that development is moving away from markdown to support more complex formatting, also as a reason that editing raw markdown is not possible (though the other mentioned reason is that collaborative editing of both wysiwyg and raw markdown is hard, which I understand).

I'm considering building a knowledge base on Outline, but need to keep the content usable in external systems (e.g. to automatically publish a website or PDF or so from a document or collection) and want to keep the option open of migrating away from Outline to another tool again in the future. In that light, I am wondering to what degree is the internal format still Markdown, and are there any thoughts to move away to a completely different format in the future?

**Comment by matthijskooijman** [comment, marked as answer]
> However, when looking at the documents.info (retrieve document) API call, the document body does look like it is still Markdown (possibly with extensions?), which is also what the [API docs suggest](https://www.getoutline.com/developers#tag/documents) ("Documents are stored in [Markdown](https://spec.commonmark.org/) formatting").

Ah, it seems this is just a markdown export for compatibility. After reading some old pull requests about switching from markdown to JSON as the canonical format, I checked my browser console and saw that the response for `documents.info` is not markdown but indeed a nested JSON description of the document. Adding the `x-api-version: 3` HTTP header to my `curl` request give me the same result, so I assume that the nested JSON format is already the current storage format and all markdown versions are actually (potentially lossy) exports.

**Reply by tommoor** [reply]
That's correct – and the reason we moved from Markdown is that it cannot represent all the many things that a modern text editor must achieve, but we'll always support a lossy Markdown output.
